ostd/src/mm/dma/mod.rs

```diff
@@ -1,72 +1,600 @@
 // SPDX-License-Identifier: MPL-2.0
 
-mod dma_coherent;
-mod dma_stream;
+//! Direct Memory Access (DMA) memory management.
+//!
+//! This module provides [`DmaCoherent`] and [`DmaStream`] abstractions for
+//! managing DMA memory regions with different remapping, caching and
+//! synchronization requirements.
+
 #[cfg(ktest)]
 mod test;
 
-use alloc::collections::BTreeSet;
-
-pub use dma_coherent::DmaCoherent;
-pub use dma_stream::{DmaDirection, DmaStream};
-use spin::Once;
+use core::{any::TypeId, fmt::Debug, marker::PhantomData, mem::ManuallyDrop, ops::Range};
 
 use super::{Daddr, Paddr};
-use crate::{arch::iommu::has_dma_remapping, mm::PAGE_SIZE, sync::SpinLock};
+use crate::{
+    arch::{
+        iommu::{self, has_dma_remapping},
+        mm::can_sync_dma,
+    },
+    cpu::{AtomicCpuSet, CpuSet},
+    error::Error,
+    mm::{
+        CachePolicy, FrameAllocOptions, HasDaddr, HasPaddr, HasPaddrRange, HasSize, Infallible,
+        PAGE_SIZE, PageFlags, PageProperty, PrivilegedPageFlags, Split, USegment, VmReader,
+        VmWriter,
+        io_util::{HasVmReaderWriter, VmReaderWriterIdentity, VmReaderWriterResult},
+        kspace::kvirt_area::KVirtArea,
+        paddr_to_vaddr,
+        page_table::vaddr_range,
+        tlb::{TlbFlushOp, TlbFlusher},
+    },
+    task::disable_preempt,
+    util::range_alloc::RangeAllocator,
+};
+#[cfg(any(debug_assertions, all(target_arch = "x86_64", feature = "cvm_guest")))]
+use crate::{sync::SpinLock, util::range_counter::RangeCounter};
+
+/// [`DmaDirection`] limits the data flow direction of [`DmaStream`] and
+/// prevents users from reading and writing to [`DmaStream`] unexpectedly.
+pub trait DmaDirection: 'static + Debug {}
+
+/// Data flows to the device.
+///
+/// From the perspective of the kernel, this memory region is writable.
+#[derive(Debug)]
+pub enum ToDevice {}
+
+impl DmaDirection for ToDevice {}
+
+/// Data flows from the device.
+///
+/// From the perspective of the kernel, this memory region is read-only.
+#[derive(Debug)]
+pub enum FromDevice {}
+
+impl DmaDirection for FromDevice {}
+
+/// Data flows both from and to the device.
+#[derive(Debug)]
+pub enum Bidirectional {}
+
+impl DmaDirection for Bidirectional {}
 
-#[derive(PartialEq)]
-pub enum DmaType {
-    Direct,
-    Iommu,
+/// A DMA memory object with coherent cache.
+#[derive(Debug)]
+pub struct DmaCoherent {
+    inner: ManuallyDrop<CoherentInner>,
+    map_daddr: Option<Daddr>,
+    is_cache_coherent: bool,
 }
 
-#[derive(Debug, PartialEq)]
-pub enum DmaError {
-    InvalidArgs,
-    AlreadyMapped,
+#[derive(Debug)]
+enum CoherentInner {
+    Segment(USegment),
+    Kva(KVirtArea, Paddr),
+}
+
+/// A DMA memory object with streaming access.
+///
+/// The kernel must synchronize the data by [`sync`] when interacting with the
+/// device.
+///
+/// [`sync`]: DmaStream::sync
+#[derive(Debug)]
+pub struct DmaStream<D: DmaDirection = Bidirectional> {
```

Comment:
Should we use `<const D: DmaDirection>` and keep `DmaDirection` as an `enum`? That would avoid the need to check it through `TypeId`.

Reply:
I just realized that this is not possible without [`adt-const-params`](https://doc.rust-lang.org/beta/unstable-book/language-features/adt-const-params.html).

As an alternative, maybe we can do this:
```
trait DmaDirection {
    const IS_TO_DEVICE: bool;
    const IS_FROM_DEVICE: bool;
}
```
(or `CAN_TO_DEVICE`/`CAN_FROM_DEVICE`)?